### PR TITLE
Switch audio source

### PR DESCRIPTION
**Pull request by skywojciechowskim**
Summary: Switch Audio source implementation
Type: Feature
Owner: Marcin Wojciechowski
Reviewers: Adam Czynszak, Stuart Pett
Coding Standard Applied: Y
Test Plan: None
Dependencies and Impacts: None
Jira: RIALTO-28

### CODE REVIEW

File: media/server/gstplayer/source/tasks/generic/AttachSource.cpp [hunk 1]
```diff
@@ -241,43 +243,97 @@ void AttachSource::execute() const
...
+                                                            &audioChangeTargetPts, &currentDispPts, &audioChangeStage,
+                                                            &caps, // may fail for amlogic - that implementation changes
+                                                                   // this parameter, it's probably used by Netflix later
+                                                            &audioAac, svpEnabled,
+                                                            m_context.streamInfo[m_attachedSource->getType()], &retVal);
+    if (!result || !retVal)
+    {
+        RIALTO_SERVER_LOG_WARN("performAudioTrackCodecChannelSwitch failed! Result: %d, retval %d", result, retVal);
+    }
+
```
[the remark this answers is not in the document]
Comment (skywojciechowskim) [reply media/server/gstplayer/source/tasks/generic/AttachSource.cpp:329]: No, I suppose. I can't find such call in Netflix code. AMLogic calls it inside performAudioTrackCodecChannelSwitch, realtek implementation doesn't call this function.

File: media/server/gstplayer/include/IGlibWrapper.h [hunk 2]
```diff
@@ -259,6 +270,16 @@ class IGlibWrapper
      * @retval A pointer to the newly-allocated copy of the memory or null in case of failure
      */
     virtual void gOnceInitLeave(gsize *location, gsize result) const = 0;
+
+    /**
+     * @brief Searches the string haystack for the last occurrence of the string needle.
+     *
+     * @param[in] haystack : A nul-terminated string.
+     * @param[in] needle   : The nul-terminated string to search for.
```
[the remark this answers is not in the document]
Comment (skywojciechowskim) [reply media/server/gstplayer/include/IGlibWrapper.h]: done

File: media/server/gstplayer/source/tasks/generic/AttachSource.cpp [hunk 3]
```diff
@@ -241,43 +243,97 @@ void AttachSource::execute() const
...
+}
+
+void AttachSource::switchAudioSource(GstCaps *caps) const
+{
+    if (m_attachedSource->getMimeType().empty())
+    {
+        RIALTO_SERVER_LOG_WARN("SKIP switching audio source. Unknown mime type");
+        return;
+    }
+    RIALTO_SERVER_LOG_INFO("Switching audio source");
```
[the remark this answers is not in the document]
Comment (skywojciechowskim) [reply media/server/gstplayer/source/tasks/generic/AttachSource.cpp]: Changed to MIL and printed new caps. Old caps are not available here.

File: media/server/gstplayer/source/GstGenericPlayer.cpp [hunk 4]
```diff
@@ -339,8 +373,9 @@ bool GstGenericPlayer::getPosition(std::int64_t &position)
 {
     // We are on main thread here, but m_context.pipeline can be used, because it's modified only in GstGenericPlayer
     // constructor and destructor. GstGenericPlayer is created/destructed on main thread, so we won't have a crash here.
-    if (!m_context.pipeline || GST_STATE(m_context.pipeline) < GST_STATE_PLAYING)
+    if (!m_context.pipeline || GST_STATE(m_context.pipeline) < GST_STATE_PAUSED)
```
Comment (skypetts) [inline media/server/gstplayer/source/GstGenericPlayer.cpp:376]: Is this an intentional change, it doesn't seem related to stream switching

Comment (skywojciechowskim) [reply media/server/gstplayer/source/GstGenericPlayer.cpp:376]: Yes, it is. performAudioTrackCodecChannelSwitch has to check the playback position in PAUSED state. According to GStreamer documentation, it allows to check current position in > READY state (so in PAUSED and PLAYING). I checked all YT tests, this change does not break anything.

File: media/client/main/include/MediaPipeline.h [hunk 5]
```diff
@@ -186,6 +186,11 @@ class MediaPipeline : public IMediaPipeline, public IMediaPipelineIpcClient, pub
      */
     std::atomic<State> m_currentState;
 
+    /**
+     * @brief The flag indicating, if audio source switch is ongoing.
+     */
+    std::atomic_bool m_audioSourceSwitchOngoing;
```
[the remark this answers is not in the document]
Comment (skypetts) [reply media/client/main/include/MediaPipeline.h:192]: As discussed we can deliver with this now

File: media/server/gstplayer/source/tasks/generic/UpdatePlaybackGroup.cpp [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        RIALTO_SERVER_LOG_DEBUG("Typefind SRC Pad Caps NULL");
+        return;
+    }
+    gchar *typefindCaps = m_gstWrapper->gstCapsToString(m_caps);
+    if (typefindCaps)
+    {
+        RIALTO_SERVER_LOG_DEBUG("Typefind SRC Pad Strm Parsed Caps %s", typefindCaps);
+        if (m_glibWrapper->gStrrstr(typefindCaps, "audio/"))
+        {
+            RIALTO_SERVER_LOG_DEBUG("Typefind Audio Caps %s", typefindCaps);
```
Comment (aczs) [inline media/server/gstplayer/source/tasks/generic/UpdatePlaybackGroup.cpp:52]: Is printing typefindCaps needed here? It was printed one log before.

File: media/server/wrappers/source/RdkGstreamerUtilsWrapper.cpp [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    }
+    rdk_gstreamer_utils::rdkGstreamerUtilsPlaybackGrp *rdkPlaybackGroup{
+        reinterpret_cast<rdk_gstreamer_utils::rdkGstreamerUtilsPlaybackGrp *>(playbackGroup)};
+    rdk_gstreamer_utils::AudioAttributes *rdkAudioAttributes{
+        reinterpret_cast<rdk_gstreamer_utils::AudioAttributes *>(audioAttr)};
+    bool result{rdk_gstreamer_utils::performAudioTrackCodecChannelSwitch(rdkPlaybackGroup, sampleAttr, rdkAudioAttributes,
+                                                                         status, ui32Delay, audioChangeTargetPts,
+                                                                         currentDispPts, audioChangeStage, appsrcCaps,
+                                                                         audioaac, svpEnabled, aSrc, ret)};
+    return result;
```
Comment (aczs) [inline media/server/wrappers/source/RdkGstreamerUtilsWrapper.cpp:60]: Why not
`return rdk_gstreamer_utils::performAudioTrackCodecChannelSwitch(rdkPlaybackGroup, sampleAttr, rdkAudioAttributes,
                                                                         status, ui32Delay, audioChangeTargetPts,
                                                                         currentDispPts, audioChangeStage, appsrcCaps,
                                                                         audioaac, svpEnabled, aSrc, ret)`?

**Review by aczs — APPROVED** [review]

File: media/server/gstplayer/source/tasks/generic/UpdatePlaybackGroup.cpp (hunk 6 above)
Comment (skywojciechowskim) [reply media/server/gstplayer/source/tasks/generic/UpdatePlaybackGroup.cpp:52]: I will remove this print in next commit

File: media/server/wrappers/source/RdkGstreamerUtilsWrapper.cpp (hunk 7 above)
Comment (skywojciechowskim) [reply media/server/wrappers/source/RdkGstreamerUtilsWrapper.cpp:60]: I will change it in next commit
